File: model-client/src/main/java/org/modelix/model/persistent/SerializationUtil.java
package org.modelix.model.persistent;



import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.net.URLDecoder;

public class SerializationUtil {
  public static String escape(String value) {
    if (value == null) {
      return "%00";
    }
    return URLEncoder.encode(value, StandardCharsets.UTF_8);
  }

  public static String unescape(String value) {
    if (value == null) {
      return null;
    }
    if ("%00".equals(value)) {
      return null;
    }
    return URLDecoder.decode(value, StandardCharsets.UTF_8);
  }

  public static String longToHex(long value) {
    return Long.toHexString(value);
  }

  public static long longFromHex(String hex) {
    return Long.parseUnsignedLong(hex, 16);
  }

  public static String intToHex(int value) {
    return Integer.toHexString(value);
  }

  public static int intFromHex(String hex) {
    return Integer.parseUnsignedInt(hex, 16);
  }

  public static String nullAsEmptyString(String str) {
    if (str == null) {
      return "";
    }
    if (str.length() == 0) {
      throw new RuntimeException("Empty string not allowed");
    }
    return str;
  }

  public static String emptyStringAsNull(String str) {
    return ((str == null || str.length() == 0) ? null : str);
  }
}

File: model-client/src/main/java/org/modelix/model/persistent/SerializationUtil.kt
package org.modelix.model.persistent

import java.net.URLDecoder
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

object SerializationUtil {
    @JvmStatic
    fun escape(value: String?): String {
        return if (value == null) {
            "%00"
        } else URLEncoder.encode(value, StandardCharsets.UTF_8)
    }

    @JvmStatic
    fun unescape(value: String?): String? {
        if (value == null) {
            return null
        }
        return if ("%00" == value) {
            null
        } else URLDecoder.decode(value, StandardCharsets.UTF_8)
    }

    @JvmStatic
    fun longToHex(value: Long): String {
        return java.lang.Long.toHexString(value)
    }

    @JvmStatic
    fun longFromHex(hex: String?): Long {
        return java.lang.Long.parseUnsignedLong(hex, 16)
    }

    @JvmStatic
    fun intToHex(value: Int): String {
        return Integer.toHexString(value)
    }

    @JvmStatic
    fun intFromHex(hex: String?): Int {
        return Integer.parseUnsignedInt(hex, 16)
    }

    @JvmStatic
    fun nullAsEmptyString(str: String?): String {
        if (str == null) {
            return ""
        }
        if (str.length == 0) {
            throw RuntimeException("Empty string not allowed")
        }
        return str
    }

    @JvmStatic
    fun emptyStringAsNull(str: String?): String? {
        return if (str == null || str.length == 0) null else str
    }
}